### ISSUE TITLE
Possible error of CRC check procedure

### ISSUE THREAD

**Issue by PandaMD-Way** [NVlabs/sionna#789]
![Image](https://github.com/user-attachments/assets/2c3486f8-3b49-4c03-8681-1353f4a4683f)

Above shortcut is the procedure in crc_decoder for CRC status check. I think the procuedure is not correct.
My understanding of the procedure is that, using the information bits to regenerate the CRC bits and compare it with the original attched bits.

Possible errors are:
1) The information bits should be variable x_info, not variable inputs which contains both original information bits and CRC bits.
2) The step, using tf.reduce_sum to sum all parity bits x_parity, does not make sense.

I think the code should be rectified as following:

x_info = inputs[...,0:-self._encoder.crc_length]
x_parity = self._encoder(x_info)[..., -self._encoder.crc_length:]
x_diff = inputs[...,-self._encoder.crc_length:] - x_parity
crc_check = tf.reduce_sum(x_diff, axis=-1, keepdims=True)

**Comment by PandaMD-Way** [comment]
Please correct me if my understanding is wrong.

**Comment by SebastianCa** [comment]
Hi @PandaMD-Way,

The way this implementation of the CRC decoder works is by re-encoding ``x_crc`` (i.e., the input sequence including the parity bits).
If ``x_crc`` is a valid codeword, the parity bits resulting from the second encoding will be `all-zero` — this is what’s being checked in the second part.

Your version should work just as well. Just one small correction in the last line: you need to take the absolute value of ``x_diff``, since it’s possible that -1 and 1 cancel out in the sum.

**Comment by SebastianCa** [comment]
Just in case you are interested, you could contribute your approach as PR as it avoids re-tracing of the encoder due to the same input dimensions.

**Comment by PandaMD-Way** [comment]
Understood, thanks for the elaboration and instruction.

--

BR,
WEI




At 2025-03-21 19:28:55, "SebastianCa" ***@***.***> wrote:

Hi @PandaMD-Way,

The way this implementation of the CRC decoder works is by re-encoding x_crc (i.e., the input sequence including the parity bits).
If x_crc is a valid codeword, the parity bits resulting from the second encoding will be all-zero — this is what’s being checked in the second part.

Your version should work just as well. Just one small correction in the last line: you need to take the absolute value of x_diff, since it’s possible that -1 and 1 cancel out in the sum.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

SebastianCa left a comment (NVlabs/sionna#789)

Hi @PandaMD-Way,

The way this implementation of the CRC decoder works is by re-encoding x_crc (i.e., the input sequence including the parity bits).
If x_crc is a valid codeword, the parity bits resulting from the second encoding will be all-zero — this is what’s being checked in the second part.

Your version should work just as well. Just one small correction in the last line: you need to take the absolute value of x_diff, since it’s possible that -1 and 1 cancel out in the sum.
